test/plugins.spec.ts

```diff
@@ -0,0 +1,153 @@
+import {
+  ICreatePlugin,
+  IGetters,
+  IModule,
+  IMutations,
+  IPlugin,
+  IState,
+  LocalStoragePlugin,
+  ILocalStoragePluginState,
+  flatten,
+  createModule,
+} from '../lib'
+
+interface IBaseTestState extends IState {
+  id: number
+}
+
+interface IOldTestState extends IBaseTestState {
+  name: string
+}
+
+interface INewTestState extends IBaseTestState {
+  age: number
+}
+
+interface ITestGetters extends IGetters {
+  isDefault: () => boolean
+}
+
+interface ITestMutations extends IMutations {
+  updateId: (id: number) => void
+}
+
+const getOldDefaultState = (): IOldTestState => ({
+  id: 0,
+  name: '',
+})
+
+const getNewDefaultState = (): INewTestState => ({
+  id: 0,
+  age: 5,
+})
+
+interface ITestModuleOptions<S extends IBaseTestState> {
+  version: number
+  stateInit: () => S
+  plugins: (ICreatePlugin<S> | IPlugin<S>)[]
+}
+
+const makeTestModule = <S extends IBaseTestState>({
+  version,
+  stateInit,
+  plugins,
+}: ITestModuleOptions<S>): IModule<S, ITestGetters, ITestMutations> =>
+  createModule<S, ITestGetters, ITestMutations>({
+    name: 'test',
+    version,
+    stateInit,
+    getters: (state) => ({
+      isDefault: () => state.id === 0 && state.name === '',
+    }),
+    mutations: (state) => ({
+      updateId: (id: number) => {
+        state.id = id
+      },
+    }),
+    plugins,
+  })
+
+describe('vue-store plugins', () => {
+  beforeEach(() => {
+    localStorage.clear()
+    jest.clearAllMocks()
+  })
+
+  describe('local stroage plugin', () => {
+    it('modify state and save with built-in local storage plugin', async () => {
+      const module = flatten(
+        makeTestModule<IOldTestState>({
+          version: 1,
+          stateInit: getOldDefaultState,
+          plugins: [LocalStoragePlugin],
+        }),
+      )
+      expect(module.id.value).toEqual(0)
+
+      // Verify values in localStorage before anything changes
+      const stringifiedStateBefore = localStorage.getItem('test')
+      expect(stringifiedStateBefore).toBeNull()
+
+      // Update id
+      const newId = 1
+      module.updateId(newId)
+      // a workaround to wait until the Vue watch-effect is done
+      await Promise.resolve()
+
+      // Verify values in localStorage after state change
+      const storedState: ILocalStoragePluginState<IOldTestState> = JSON.parse(
+        localStorage.getItem('test') ?? '',
+      )
+      expect(storedState.state.id).toBe(1)
+      expect(storedState.__version).toBe(1)
+    })
+
+    it('migrate states when version changes', async () => {
```

Comment:
This test case is kind of difficult to read and understand.
The intent is to verify if the migrate feature really works in LocalStoragePlugin.

In order to test this, I have to add different interfaces for different states (`IOldTestState` and `INewTestState`), and two functions to generate default states based on those interfaces.

In addition to verifying in-memory states, we also have to verify the states being written in localStorage.
- Properties that only exist in the old state should not show up in the new in-memory state and localStorage.
- Properties that exist in both old and new states should not be reset back to the new default value.
- Properties that only exist in the new state should show up in the new in-memory state and localStorage.